### DISCUSSION TITLE
Type error on custom key

### DISCUSSION THREAD

**Discussion by lsarni** [urql-graphql/urql#3651] [Q&A] [answered]
We are defining a custom key as it is [documented here](https://commerce.nearform.com/open-source/urql/docs/graphcache/normalized-caching/#custom-keys-and-non-keyable-entities):
```
offlineExchange({
  keys: {
    Item: data => data.slug,
  },
})
```

But we get a type error:
```
Type 'DataField' is not assignable to type 'string | null'.
  Type 'number' is not assignable to type 'string'.ts(2322)
urql-exchange-graphcache-chunk.d.ts(793, 5): The expected type comes from the return type of this signature.
```

The cache works as expected though, and it doesn't if we remove this custom key. Here someone brought up the same problem https://github.com/urql-graphql/urql/discussions/2172 and what they did as a workaround still works, but is there another solution?

**Comment by kitten** [comment, marked as answer]
No, without typegen there's no way of knowing what the scalar value of the field may be. Hence, it's typed more generically, and if you know the type you should cast it to indicate that it's safe.

In general, these are the kind of cases where typegen is really useful: https://www.npmjs.com/package/@graphql-codegen/typescript-urql-graphcache
However, we haven't replaced the GraphQL Codegen plugin with our own implementation, e.g. in `gql.tada` yet.
